h/settings.py

```diff
@@ -13,54 +14,76 @@ class SettingError(Exception):
-
-    def __init__(self, setting, message):
-        self.setting = setting
-        self.message = message
-
-        # Test seam
-        self.warn = log.warn
-
-    def __call__(self, environ):
-        result = self.setting(environ)
-        if result is not None:
-            self.warn(self.warning)
-        return result
-
-    @property
-    def warning(self):
-        return 'use of {s} is deprecated: {m}'.format(s=self.setting,
-                                                      m=self.message)
-
-
-class EnvSetting(object):
-
-    """An (optionally typed) setting read from an environment variable."""
-
-    def __init__(self, setting, varname, type=None):
-        self.setting = setting
-        self.varname = varname
-        if type is not None:
-            self.type = type
-        else:
-            self.type = str
-
-    def __call__(self, environ):
-        if self.varname in environ:
+class SettingsManager(object):
+    def __init__(self, settings=None, environ=None):
+        self.settings = settings or {}
+        self._environ = environ or os.environ
+
+    def set(self,  # pylint: disable=too-many-arguments
+            name,
+            envvar,
+            type_=str,
+            required=False,
+            default=None,
+            deprecated_msg=None):
+        """
+        Set settings[name] to envvar or default.
+
+        Environment variables override config file settings and
+        defaults: if the named envvar is in the environment then set
+        settings[name] to the value of the envvar.
+
+        Config file settings override defauits: if the envvar isn't
+        in the environment, and the named setting is already present
+        in settings, then leave settings[name] unchanged.
+
+        Defaults are used if the setting isn't in the environment or
+        the config file: if envvar isn't in the environment and name
+        isn't in settings then set settings[name] to the given default
+        value.
+
+        If envvar isn't in the environment and name isn't in settings
+        and no default is given then leave settings[name] unset. If
+        this happens and required=True then raise SettingError.
+
+        If deprecation_msg is not None and the envvar is in the
+        environment then log a warning envvar deprecation message.
+
+        :param name: the name of the pyramid config setting
+        :type name: str
+        :param envvar: the environment variable name
+        :type envvar: str
+        :param type_: the type to type cast the envvar or default value
+        :param required: True if the the pyramid config setting is required
+        :type required: bool
+        :param default: a default value to use if the envvar isn't set
+        :param deprecated_msg: a deprecated envvar setting message to display
+        :type deprecated_msg: str
+        :raises SettingsError: if required and not set or type casting fails
```

Comment:
If there are two different reasons why it might raise `SettingsError` then I think you can put two `:raises SettingsError: ...` lines. That gives you more space to describe each reason, so you don't have to cram them into one line:

```
:raises SettingsError: if required=True and the setting isn't found in either environ or settings
:raises SettingsError: if type-casting the setting's value fails
```